sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -0,0 +1,179 @@
+    public static BinaryContent Create(BinaryData value) => new BinaryDataBinaryContent(value.ToMemory());
+
+    /// <summary>
+    /// Creates an instance of <see cref="BinaryContent"/> that wraps a <see cref="IPersistableModel{T}"/>.
+    /// </summary>
+    /// <param name="model">The <see cref="IPersistableModel{T}"/> to write.</param>
+    /// <param name="options">The <see cref="ModelReaderWriterOptions"/> to use.</param>
+    /// <returns>An instance of <see cref="BinaryContent"/> that wraps a <see cref="IPersistableModel{T}"/>.</returns>
+    public static BinaryContent Create<T>(T model, ModelReaderWriterOptions? options = default) where T : IPersistableModel<T>
+        => new ModelBinaryContent<T>(model, options ?? ModelWriteWireOptions);
+
+    /// <summary>
+    /// Attempts to compute the length of the underlying body content, if available.
+    /// </summary>
+    /// <param name="length">The length of the underlying data.</param>
+    public abstract bool TryComputeLength(out long length);
+
+    /// <summary>
+    /// Writes contents of this object to an instance of <see cref="Stream"/>.
+    /// </summary>
+    /// <param name="stream">The stream to write to.</param>
+    /// <param name="cancellationToken">To cancellation token to use.</param>
+    public abstract Task WriteToAsync(Stream stream, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Writes contents of this object to an instance of <see cref="Stream"/>.
+    /// </summary>
+    /// <param name="stream">The stream to write to.</param>
+    /// <param name="cancellationToken">To cancellation token to use.</param>
+    public abstract void WriteTo(Stream stream, CancellationToken cancellationToken);
+
+    /// <inheritdoc/>
+    public abstract void Dispose();
+
+    private sealed class BinaryDataBinaryContent : BinaryContent
+    {
+        private readonly ReadOnlyMemory<byte> _bytes;
+
+        public BinaryDataBinaryContent(ReadOnlyMemory<byte> bytes)
+        {
+            _bytes = bytes;
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            length = _bytes.Length;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellation)
+        {
+            byte[] buffer = _bytes.ToArray();
+            stream.Write(buffer, 0, buffer.Length);
+        }
+
+        public override async Task WriteToAsync(Stream stream, CancellationToken cancellation)
+            => await stream.WriteAsync(_bytes, cancellation).ConfigureAwait(false);
+
+        public override void Dispose() { }
+    }
+
+    private sealed class ModelBinaryContent<T> : BinaryContent where T : IPersistableModel<T>
+    {
+        private readonly T _model;
+        private readonly ModelReaderWriterOptions _options;
+
+        // Used when _model is an IJsonModel
+        private ModelWriter? _writer;
```

Comment:
Why do we need a field for the writer? Why not create the writer when we need it, use it, and then get rid of it.

Reply:
I think this is the discussion you have been having with @m-nash about whether this type can feasibly not implement IDisposable.  @m-nash has said we have to keep it so we don't recreate it every time it is needed.

Reply:
I don't think this is the same discussion. We cannot remove IDisposable because we borrow buffers from the pool. But we don't need the writer after we write the buffers.

Reply:
Tracking resolution with https://github.com/Azure/azure-sdk-for-net/issues/41199